Add support to Android using Kotlin code generation

husseinala opened the pull request:
Resolves #1 

# Description

Where I work, we've been using Arkana on our iOS projects and we're big fans of the project, so would like to adopt it for our Android projects too, which is why we thought it would be helpful to kick off the process of adding Kotlin support.

This PR adds a `KotlinCodeGenerator` and modifies the execution entry point to determine which generator to run based on the specified `--lang` argument flag.

The following changes have also been made:

### Arkana Arguments

- Added a `--lang` flag used to specify the generator to use for e.g. `kotlin`. Defaults to `swift`.

### Arkana YAML config

- Added a `kotlin_package_name` option which is used by the Kotlin generator to determine the package name of the generated module. Defaults to `com.arkanakeys`.
- Added a `kotlin_sources_path` option which is used by the Kotlin generator to determine the path for generated Kotlin classes. Defaults to src/main/kotlin.
- Added a `should_generate_gradle_build_file` option which is useful if your `result_path` is a path of an existing module and you would like to avoid generating a new `build.gradle` file.

# Generated module structure

<img width="410" alt="Screenshot 2023-12-05 at 11 14 59" src="https://github.com/rogerluan/arkana/assets/1149260/a2f2a799-c328-4a60-89d8-2ad00be453e6">

# Generated sources

### {namespace}Environment.kt

Contains interface declaration for the environment-specific keys.
<img width="689" alt="Screenshot 2023-12-05 at 11 17 53" src="https://github.com/rogerluan/arkana/assets/1149260/520d9116-b7c4-4cd0-b2a3-7c1080139fce">

### {namespace}.kt

Contains the Keys object:

<img width="1245" alt="Screenshot 2023-12-05 at 11 19 58" src="https://github.com/rogerluan/arkana/assets/1149260/3776cb4a-d31e-4c3d-93d0-f459ea663181">

husseinala commented:
Thank you @rogerluan! Looking forward to the feedback from everyone 😊

JARMourato reviewed (commented):
Its missing the README adjustments 😉

marceltex commented on lib/arkana.rb:
```diff
@@ -43,7 +44,14 @@ def self.run(arguments)
+    generator = case config.current_lang.downcase
+      when "swift" then SwiftCodeGenerator
+      when "kotlin" then KotlinCodeGenerator
+      else UI.crash("Unknown output lang selected: #{config.current_lang.downcase}")
```
Do we need to change `current_lang` to lower case when it is displayed in the error message? 

It's a small detail, but IMO it would be slightly easier for the user to identify their mistake if we just printed the argument as they provided it, without changing the case of the letters.

```suggestion
      else UI.crash("Unknown output lang selected: #{config.current_lang}")
```

marceltex commented on lib/arkana/helpers/kotlin_template_helper.rb:
```diff
@@ -0,0 +1,22 @@
+      when :string then "decode"
+      when :boolean then "decodeBoolean"
+      when :integer then "decodeInt"
+      else raise "Unknown variable type '#{type}' received.'"
```
Looks like there is an additional single quote in this string. Is it necessary?

```suggestion
      else raise "Unknown variable type '#{type}' received."
```

rogerluan commented on lib/arkana/models/arguments.rb:
```diff
@@ -33,6 +36,9 @@ def initialize
       opt.on("-i", "--include-environments debug,release", "Optionally pass the environments that you want Arkana to generate secrets for. Useful if you only want to build a certain environment, e.g. just Debug in local machines, while only building Staging and Release in CI. Separate the keys using a comma, without spaces. When ommitted, Arkana generate secrets for all environments.") do |o|
         @include_environments = o.split(",")
       end
+      opt.on("-l", "--lang kotlin", "Language to produce keys for, for e.g. kotlin. See the README for more information") do |o|
```
```suggestion
      opt.on("-l", "--lang kotlin", "Language to produce keys for, e.g. kotlin, swift. Defaults to swift. See the README for more information") do |o|
```

rogerluan commented on lib/arkana/kotlin_code_generator.rb:
```diff
@@ -0,0 +1,54 @@
+require "fileutils"
+require_relative "helpers/string"
+
+# Responsible for generating Kotlin source and test files.
```
This generator isn't generating any test files 🙈 

Instead of updating the comment, could you implement unit test files like we're doing for Swift? 🙇 

They are kotlin unit test files generated when the source files are generated, that make sure the encoding/decoding is working as expected

husseinala replied:
Yep, of course, will work on getting the tests generated too.

rogerluan reviewed (changes requested):
> [!NOTE]
> Partial review for now!

I left some minor comments so far, as I haven't reviewed the core logic of the PR. Some things I noticed (also piggy backing on some of the comments that were already left above by other reviewers):

1. We should be updating the README as well to reflect all the changes (e.g. remove the "coming soon" section, adding proper usage documentation for the new kotlin features, perhaps add some more instructions on how to adequate your project for kotlin generation, etc).
2. Kotlin unit tests are super valuable as they guarantee the actual encoding/decoding is working as expected, and that it doesn't regress in the future :) could you add those?
3. Do the generated kotlin files pass a standard kotlin linter out-of-the-box? One with a "widely adopted" style guide 😇 
4. Last but not least, could you also implement an e2e test to be run in CI via Rakefile, like we're doing with Swift? These references will be helpful:
    - https://github.com/rogerluan/arkana/blob/main/.github/workflows/swift-tests.yml
    - https://github.com/rogerluan/arkana/blob/2bf88668fc970e313437e99a66c815055dabeecc/Rakefile#L15L22

Looking forward to continue reviewing this PR when I have some extra time! It's looking promising! 🙌

husseinala commented on template.yml:
```diff
@@ -9,6 +9,9 @@ swift_declaration_strategy: let # Optional. One of: lazy var, var, let. Defaults
 should_generate_unit_tests: true # Optional. One of: true, false. Defaults to true.
 should_cocoapods_cross_import_modules: true # Optional. If false, the two generated cocoapods pods do not cross import (for manual import of the Sources folders), useful when building SDKs, for instance. One of: true, false. Defaults to true.
 package_manager: cocoapods # Optional. This setting defines which type of dependency management you're going ot use to integrate Arkana into your project. One of: spm, cocoapods. If you use both in your project, use cocoapods here. Defaults to spm.
+kotlin_package_name: 'com.tmrw.arkanakeys' # Optional. used by the Kotlin generator to determine the package name of the generated module. Defaults to com.arkanakeys.
```
This slipped from testing, will update it to `com.arkanakeys`.

husseinala commented:
> Note
> 
> Partial review for now!
> 
> I left some minor comments so far, as I haven't reviewed the core logic of the PR. Some things I noticed (also piggy backing on some of the comments that were already left above by other reviewers):
> 
> 1. We should be updating the README as well to reflect all the changes (e.g. remove the "coming soon" section, adding proper usage documentation for the new kotlin features, perhaps add some more instructions on how to adequate your project for kotlin generation, etc).
> 2. Kotlin unit tests are super valuable as they guarantee the actual encoding/decoding is working as expected, and that it doesn't regress in the future :) could you add those?
> 3. Do the generated kotlin files pass a standard kotlin linter out-of-the-box? One with a "widely adopted" style guide 😇
> 4. Last but not least, could you also implement an e2e test to be run in CI via Rakefile, like we're doing with Swift? These references will be helpful:
>    
>    * https://github.com/rogerluan/arkana/blob/main/.github/workflows/swift-tests.yml
>    * https://github.com/rogerluan/arkana/blob/2bf88668fc970e313437e99a66c815055dabeecc/Rakefile#L15L22
> 
> Looking forward to continue reviewing this PR when I have some extra time! It's looking promising! 🙌

Thanks @JARMourato, @marceltex & @rogerluan for taking the time to go through the PR 🙏🏼 , will work on addressing all the raised issues and push all the necessary changes, including making sure all the generated code passes Kotlin's official [codeing conventions](https://kotlinlang.org/docs/coding-conventions.html)

rogerluan commented on template.yml:
```diff
@@ -9,6 +9,9 @@ swift_declaration_strategy: let # Optional. One of: lazy var, var, let. Defaults
 should_cocoapods_cross_import_modules: true # Optional. If false, the two generated cocoapods pods do not cross import (for manual import of the Sources folders), useful when building SDKs, for instance. One of: true, false. Defaults to true.
 package_manager: cocoapods # Optional. This setting defines which type of dependency management you're going ot use to integrate Arkana into your project. One of: spm, cocoapods. If you use both in your project, use cocoapods here. Defaults to spm.
+kotlin_package_name: 'com.tmrw.arkanakeys' # Optional. used by the Kotlin generator to determine the package name of the generated module. Defaults to com.arkanakeys.
+kotlin_sources_path: 'src/main/java' # Optional. used by the Kotlin generator to determine the path for generated kotlin classes. Defaults to src/main/kotlin.
```
Ah gotcha, we can keep it src/main/java then :) makes sense now

husseinala commented:
Hi @rogerluan, sorry for the delay, I've been a bit busy the last two weeks, but I finally got around to making the requested changes which address the following points:

1. The readme has been updated to include information about Android/Kotlin support.
2. Updated the Kotlin code generator to generate Kotlin unit tests when `should_generate_gradle_build_file` is set to `true`.
3. Updated the Rakefile to include a `test_kotlin` task. This required the addition of a couple of support files which I included inside `spec/fixtures/kotlin` to be able to run the Kotlin tests outside a normal Android project.
4. Updated the Kotlin code template to ensure the generated code is in line with the official JetBrains Kotlin code style.

rogerluan commented:
I made some changes to the Kotlin erb templates to adjust the code style/formatting, more specifically the line breaks 😊 I think they look better this way. Here's a visual diff of the changes if you're curious:

<details><summary>Click here to expand</summary>
<p>

<img width="1093" alt="image" src="https://github.com/rogerluan/arkana/assets/8419048/af1a97f2-9c8a-4729-a643-55942b35d230">
<img width="1093" alt="image" src="https://github.com/rogerluan/arkana/assets/8419048/03f1f50c-5643-4bff-a3b0-84fbdd2534f6">
<img width="1096" alt="image" src="https://github.com/rogerluan/arkana/assets/8419048/8a086720-d930-4dae-8c48-e8c4c13f0e2e">
<img width="1053" alt="image" src="https://github.com/rogerluan/arkana/assets/8419048/89fa1315-c991-4538-9a5c-97d0ae1d41d6">

</p>
</details>

rogerluan commented on lib/arkana/templates/kotlin/arkana.kt.erb:
```diff
@@ -0,0 +1,59 @@
+
+    object Global {
+<% @global_secrets.each_with_index do |secret, index| %>
+        val <%= secret.key.camel_case %>: <%= KotlinTemplateHelper.kotlin_type(secret.type) %>
```
In Swift I decided to allow users to decide which type of variable declaration they'd like to use (`lazy var` - a lazily instantiated variable, `var` - a variable, `let` - a constant). Do you think this would also be valuable here?

No need to make any changes in this PR regarding this, just wondering if in Android it would make any sense to use other types of variable declarations, for potential future improvements :)

humblerookie replied:
I think that might be an overkill here, lazy evaluation adds more overhead in kotlin for trivial evaluations. It may not be appropriate for this use case.

husseinala replied:
We can always allow the user the option to enable lazy properties in the future but as @humblerookie has mentioned, `lazy` properties in Kotlin use [delegates](https://kotlinlang.org/docs/delegated-properties.html), which could potentially add more overhead than repeated calls to the `decode` methods.

rogerluan replied:
That's good to know! I trust your experience 💪 thanks for the input! ❤️

rogerluan commented on lib/arkana/templates/kotlin/arkana_tests.kt.erb:
```diff
@@ -0,0 +1,138 @@
+        assertEquals(globalSecrets.secretWithWeirdCharactersKey, "` ~ ! @ # % ^ & * ( ) _ - + = { [ } } | : ; ' < , > . ? /")
+    }
+<% end %>
+}
```
FTFY 😊

rogerluan commented on spec/fixtures/kotlin/build.gradle.kts:
```diff
@@ -0,0 +1,22 @@
+
+kotlin {
+    jvmToolchain(11)
+}
```
FTFY 😊

rogerluan commented on README.md:
````diff
@@ -37,29 +37,73 @@
+Usage using the example code above:
 
-<details><summary>Click here to show an image with the preview!</summary>
+```kotlin
+import com.arkanakeys.MySecrets
+
+// Designed with testability and DI in mind
+println(MySecrets.Global.someBooleanSecret)
+println(MySecrets.Global.someIntSecret)
+println(MySecrets.Global.mySecretAPIKey)
+
+// Simulating environment selection using a random boolean value
+val keys = if (Math.random() < 0.5) MySecrets.Dev else MySecrets.Staging
+println(keys.serviceKey)
+```
````
@husseinala could you confirm whether this implementation is valid? It's just a sample code, but I mean, I asked ChatGPT to generate this lol I don't even know if it's valid syntax 😂

husseinala replied:
Yep, it's valid and looks great! ChatGPT did a good job 😄

rogerluan replied:
haha sweet!! Thanks for reviewing it 😁

rogerluan reviewed (commented):
I finished reviewing this PR and addressing all the nitpicks I had in mind 🤓 

Solid PR @husseinala you did an awesome work here! Thank you so much!

There are just a couple things left that I asked you to review, whenever you have some time 🙇 other than that, just left some misc questions 😊 

Looking forward to getting this merged! 🚀🚀

husseinala commented on spec/fixtures/kotlin/gradlew:
So these files are required by the [Gradle Build Tool](https://kotlinlang.org/docs/gradle.html) which is responsible for downloading the required dependencies (Kotlin Compiler, [Kotlin Standard Library](https://kotlinlang.org/api/latest/jvm/stdlib/), and [Kotlin Test Libaray](https://kotlinlang.org/api/latest/kotlin.test/)), compiling the generated code, and running the tests. We could do all of that manually without using Gradle, but then it will be a bit more complicated and probably require other support files.

rogerluan replied:
Thanks for explaining @husseinala! 🙏 Sounds good to me 😊

rogerluan reviewed (approved):
I've merged latest master into this branch, this is now good to go!! 🚀 

Thanks once again for your contribution @husseinala! ❤️

husseinala commented:
Yaay 🎉 thank you for taking the time to review and clean up the PR @rogerluan. Really glad I was able to contribute to this project! ❤️
